### DISCUSSION TITLE
Galaxy S22 Ultra - xDrip Bluetooth connection issue

### DISCUSSION THREAD

**Discussion by gyusunglee** [NightscoutFoundation/xDrip#2048] [Q&A]
I was using xDrip on my Galaxy Note 10+ over the last 2,5 years. After recently switching to Galaxy S22 Ultra, I have been using the 2020.11.27 version for a week and then decided to update to the 2021.10.04 version. As soon as I update the xDrip, unfortunately, the Bluetooth kept disconnecting after some reading. once I had Bluetooth paired, I got a few readings. However, soon after that, my phone asked if I wanted to pair the transmitter. Once paired, I again got several readings. Over the last one hour, I had to pair my Bluetooth three times. The interval between the pairing varied.

So, I decided to go back to the old version of xDrip. Interestingly, the Bluetooth problem was gone. I had no problem for several days. I then decided to try the newest version again. Well. The pairing issues are happening again. So I am going back to the old version. Navid Fo suggested that I report this problem here. Please let me know what I can do. Thanks.

I am attaching some screenshots from my xDrip. Please let me know if you need any other information. Thanks.

<img width="350" src="https://user-images.githubusercontent.com/50647359/158096795-20e7e179-40eb-410b-8c0e-f599440c5c86.jpg" /> <img width="350" src="https://user-images.githubusercontent.com/50647359/158096810-2895d899-2d20-480f-a0ba-70fd1cc78a30.jpg" /> <img width="350" src="https://user-images.githubusercontent.com/50647359/158096819-bb74c594-8cd3-48a1-a8d6-10018cd289a8.jpg" /> <img width="350" src="https://user-images.githubusercontent.com/50647359/158096822-f52ab0d1-77da-42cc-9eeb-b62484fe3fab.jpg" />

**Comment by jamorham** [comment]
Try turning the "allow ob1 unbonding" off. Samsung devices have a bug where the pairing state is occasionally incorrectly reported by the android framework.

**Reply by tolot27** [reply]
> Samsung devices have a bug where the pairing state is occasionally incorrectly reported by the android framework.

Which Samsung devices/Android versions still have this bug? I assume this option was introduced some years ago for some specific Samsung devices running (now) older Android versions. Nowadays, this should be fixed. Also, there is no such option in other applications. IMHO, the option "Samsung workaround" should handle this transparently.

**Reply by jamorham** [reply]
I've seen various bugs with how samsung handles pairing data on different samsung phones including flagship and with latest available roms. As far as I know it is less of an issue with older models.  I've never seen it on Android 8. I think it is something to do with the pairing information now being stored within trustzone and not always available.

For example, if you create some code to enumerate all bonded devices using `mBluetoothAdapter.getBondedDevices()` every 5 minutes and check for a specific bonded device. What you'll see on phones like Samsung S9, Samsung S10, Samsung S21 is that occasionally the device wont appear in that list and during that time it wont be possible to create a bonded connection to the device. You can re-run the test a second later and it will appear, other times it is gone for a little longer. Sometimes the pairing disappears permanently.  As as result of this I've had to put some very awkward code in to the Insight driver within APS to be able to detect permanently lost pairings but not to get confused by temporarily lost pairings.

When the S21 came out we started getting reports of lost pairing and the debug information showed that the transmitter reported that it had lost the pairing. I'm not sure what we could be doing to cause this and the fact that APS experiences similar lost pairing issues and I've only seen it on samsung makes me think that is where the problem is. I'm happy to investigate all possible causes though especially as OP reports that different versions behave differently.

**Comment by gyusunglee** [comment]
Thanks. Someone told me this, and I tried. While I no longer get Bluetooth pairing messages, I get a few readings and no readings for 15-20 minutes. While the xDrip system status says connected, data does not show up (missing data). Thanks.

**Reply by jamorham** [reply]
Can you confirm whether you start getting data again later automatically or whether after readings stop that you don't get any more?

**Comment by Navid200** [comment]
Try the January 29, 2021 release:  https://github.com/NightscoutFoundation/xDrip/releases/tag/2021.01.29

Does it also provide connectivity the same as the November 27, 2020 release for you?

**Reply by gyusunglee** [reply]
Navid. Thanks for your suggestion. The newest version seems to work with my phone. I got no pairing messages any longer, and I do not see any missing data. After confirming this, re-enabled "allow ob1 unbending" option just for my curiosity. It still works.  In addition to the Bluetooth issue, the battery drain issue seems to be resolved as well. Thanks a lot.

**Comment by Navid200** [comment]
It could be very useful to know two releases in sequence one with and the other without the problem.  Then, we know which build could have caused the problem and focus on it.

**Reply by jamorham** [reply]
If OP has the patience to do this then it would be very interesting to try different versions to see if we can pinpoint any specific change that could help in resolving the problem overall.  Equally I would like them to try the latest nightly to confirm the problem still exists there.

**Reply by gyusunglee** [reply]
Unfortunately, I am extremely busy by the end of this month, so I will not be able to do this in a few days. However, I am willing to help you guys to understand better what caused the issue. I will need more explanation regarding what you want me to do.

**Reply by Navid200** [reply]
You said: 
November 27, 2020 was working fine.  After switching to October 4, 2021 and with your original settings, you had intermittent connectivity, which is not good of course.
We need you to try different versions between November 27, 2020 and October 4, 2021, with your original settings, to identify the first version that caused problems.

You can perform a binary search.  The period between Nov 27, 2020 and Oct 4, 2021 is about 10 months.  Dividing that by 2, you have 5 months.  Add 5 months to November 27 2020 and you are in April 2021.  You can try https://github.com/NightscoutFoundation/xDrip/releases/tag/2021.04.24.

If that version also results in intermittent connectivity, the target is older.  If that version works fine as November 27, 2020 did, the target we are after is younger.  So, depending on the result, the next step will be positive or negative.  And the next step size will also have to be smaller.  So, you can make it a 2-month step size.

If you keep doing this, after 5-6 steps, you will find two releases one right after the other with one working fine and the other breaking.  If you tell us what those two releases are, it will be very helpful.

**Reply by jamorham** [reply]
Please try latest nightly first in case the issue is already resolved.
